internal/yunikorn/sync.go

```diff
@@ -94,55 +95,117 @@ func (s *Service) upsertPartitions(ctx context.Context) ([]*dao.PartitionInfo, e
 	return partitions, nil
 }
 
-// upsertPartitionQueues fetches queues for each partition and upserts them into the database
-func (s *Service) upsertPartitionQueues(ctx context.Context, partitions []*dao.PartitionInfo) ([]*dao.PartitionQueueDAOInfo, error) {
+// syncQueues fetches queues for each partition and upserts them into the database
+func (s *Service) syncQueues(ctx context.Context, partitions []*dao.PartitionInfo) ([]*dao.PartitionQueueDAOInfo, error) {
 	logger := log.FromContext(ctx)
 
-	// Create a wait group as a separate goroutine will be spawned for each partition
-	wg := sync.WaitGroup{}
-	wg.Add(len(partitions))
+	var errs []error
+	var allQueues []*dao.PartitionQueueDAOInfo
+	var mu sync.Mutex
+	var wg sync.WaitGroup
 
-	// Protect the queues and errors slices with a mutex so multiple goroutines can safely append queues
-	mutex := sync.Mutex{}
+	addErr := func(err error) {
+		mu.Lock()
+		defer mu.Unlock()
+		errs = append(errs, err)
+	}
 
-	var errs []error
-	var queues []*dao.PartitionQueueDAOInfo
+	addQueues := func(queues []*dao.PartitionQueueDAOInfo) {
+		mu.Lock()
+		defer mu.Unlock()
+		allQueues = append(allQueues, queues...)
+	}
+
+	addJob := func(partition *dao.PartitionInfo) {
+		wg.Add(1)
```

Comment:
wg.Add should never be within a function. This creates a hard-to-find race condition where, if this goroutine is not yet started, and others signal with wg.Done, the counter can reach 0 without having this function executed.